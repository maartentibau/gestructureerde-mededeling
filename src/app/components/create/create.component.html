<ogm-number [ogm]="ogm()?.numberFormat ?? null"></ogm-number>

<ogm-input
  (ogmInputChange)="ogmInputChangeHandler($event)"
  placeholderMessage="vul een getal in van maximaal 10 cijfers"
></ogm-input>

@if (ogm()) {
  <ogm-controls
    [ogm]="ogm()"
    [copyNumber]="true"
    [copyOgm]="true"
    (copyNumberClick)="copyNumberClickHandler($event)"
    (copyOgmClick)="copyOgmClickHandler($event)"
  ></ogm-controls>
}
