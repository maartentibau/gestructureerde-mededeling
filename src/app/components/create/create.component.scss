.ogm-form {
  text-align: center;
  padding: 0 10%;
}

input {
  text-align: center;
  width: 100%;
  margin: 1em 0;
  padding: 8px 24px;
  border-radius: 4px;
  border: 1px solid #e9e9e9;
  font-size: 2.5rem;
  box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14),
    0px 1px 3px 0px rgba(0, 0, 0, 0.12);
  outline: none;
  color: #103183;

  &::placeholder {
    color: #e1e1e1;
  }
}
