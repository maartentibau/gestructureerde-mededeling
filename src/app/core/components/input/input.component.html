<div class="ogm-form">
  <input
    type="text"
    [attr.maxlength]="validate ? 12 : 10"
    [formControl]="ogmInput"
    [attr.placeholder]="placeholderMessage"
  />
</div>
