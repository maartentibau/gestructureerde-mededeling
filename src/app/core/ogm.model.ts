export interface OgmData {
  number: string | null;
  numberFormat: string | null;
}
