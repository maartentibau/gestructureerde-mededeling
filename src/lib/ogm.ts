export function ogm(): string {
  return 'ogm';
}
