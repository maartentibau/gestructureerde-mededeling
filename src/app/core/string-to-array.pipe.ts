import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
