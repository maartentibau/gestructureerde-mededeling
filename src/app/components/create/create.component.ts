import { ChangeDetectionStrategy, Component, inject, signal } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MatSnackBar, MatSnackBarModule } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { ControlsComponent } from '../../core/components/controls/controls.component';
import { InputComponent, Ogm } from '../../core/components/input/input.component';
import { NumberComponent } from '../../core/components/number/number.component';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmData } from '../../core/ogm.model';
import { OgmService } from '../../core/services/ogm.service';

@Component({
  imports: [NumberComponent, InputComponent, ControlsComponent, MatSnackBarModule, FormsModule],
  selector: 'ogm-create',
  templateUrl: './create.component.html',
  styleUrls: ['./create.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CreateComponent {
  #ogmService: OgmService = inject(OgmService);
  #snackBar: MatSnackBar = inject(MatSnackBar);
  #title: Title = inject(Title);

  readonly ogm = signal<OgmData | null>({ number: null, numberFormat: this.#ogmService.init() });

  constructor() {
    this.#title.setTitle(`${DEFAULT_TITLE} - Zelf een gestructureerde mededeling maken`);
  }

  handleOgmChange({ ogm }: Ogm) {
    const cleanOgm = this.#ogmService.clean(ogm ?? '');

    this.ogm.set(
      cleanOgm === ''
        ? { number: null, numberFormat: this.#ogmService.init() }
        : { number: cleanOgm, numberFormat: ogm },
    );
  }

  copyNumberClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  copyOgmClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.#snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
