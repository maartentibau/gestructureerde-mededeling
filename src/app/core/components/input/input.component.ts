import { Component, EventEmitter, inject, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { ReactiveFormsModule, UntypedFormControl } from '@angular/forms';
import { Observable, Subject } from 'rxjs';
import { filter, map, takeUntil, tap } from 'rxjs/operators';
import { OgmService } from '../../services/ogm.service';

export interface OgmInputChange {
  ogm: string;
  isValid: boolean | null;
}

@Component({
  standalone: true,
  imports: [ReactiveFormsModule],
  selector: 'ogm-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.scss'],
})
export class InputComponent implements OnInit, OnDestroy {
  #ogmService: OgmService = inject(OgmService);

  @Input() validate!: boolean;
  @Input() placeholderMessage!: string;

  @Output() readonly ogmInputChange: EventEmitter<OgmInputChange>;

  readonly ogmInput: UntypedFormControl;
  readonly destroy$: Subject<void>;

  ogm$!: Observable<string>;

  constructor() {
    this.ogmInputChange = new EventEmitter<OgmInputChange>();

    this.ogmInput = new UntypedFormControl(null);
    this.destroy$ = new Subject<void>();
  }

  ngOnInit() {
    const ogmInitValue: string = this.#ogmService.init();
    const hasSpaces: RegExp = new RegExp('[\\s]');

    const omgInputChanges$ = this.ogmInput.valueChanges.pipe(
      filter((value) => {
        if (isNaN(Number(value)) || hasSpaces.test(value)) {
          const cleanValue = this.#ogmService.clean(value);
          this.ogmInput.setValue(cleanValue, { emitEvent: false });
          return false;
        }

        return true;
      })
    );

    if (this.validate) {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.#ogmService.format(value.toString().padEnd(12, ' ')) : ogmInitValue))
      );
    } else {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.#ogmService.generate(value, true) : ogmInitValue))
      );
    }

    this.ogm$
      .pipe(
        tap((ogm: string) => {
          let isValid = null;

          if (this.validate) {
            const ogmNumber = this.#ogmService.clean(ogm);
            isValid = ogmNumber.length === 12 ? this.#ogmService.validate(ogm) : null;
          }

          this.ogmInputChange.emit({ ogm, isValid });
        }),
        takeUntil(this.destroy$)
      )
      .subscribe();
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
