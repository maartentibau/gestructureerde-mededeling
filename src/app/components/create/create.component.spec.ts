import { AsyncPipe, JsonPipe } from '@angular/common';
import { Component, input } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { render } from '@testing-library/angular';
import { Ogm } from '../../core/components/input/input.component';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmData } from '../../core/ogm.model';
import { OGM_EMPTY } from '../../core/services/ogm.service';

import { CreateComponent } from './create.component';

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'ogm-number',
  template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `,
})
class MockNumberComponent {
  ogm = input<string | null>(null);
  isValid = input<boolean | null>(null);
}

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'ogm-input',
  template: `
    <div>{{ validate | json }}</div>
    <div>{{ placeholderMessage | json }}</div>
  `,
})
class MockInputComponent {
  validate = input<boolean>(false);
  placeholderMessage = input<string>('');
}

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'ogm-controls',
  template: `
    <div>{{ refresh | json }}</div>
    <div>{{ copyNumber | json }}</div>
    <div>{{ copyOgm | json }}</div>
    <div>{{ ogm | json }}</div>
  `,
})
class MockControlsComponent {
  refresh = input<boolean>(false);
  copyNumber = input<boolean>(false);
  copyOgm = input<boolean>(false);
  ogm = input<OgmData | null>(null);
}

describe('CreateComponent', () => {
  let component: CreateComponent;
  let fixture: ComponentFixture<CreateComponent>;
  let titleService: Title;

  beforeEach(async () => {
    ({ fixture, component, titleService } = await setup());
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('set the correct HTML title', () => {
      // prepare
      const title: string = `${DEFAULT_TITLE} - Zelf een gestructureerde mededeling maken`;
      jest.spyOn(titleService, 'setTitle');

      // act
      TestBed.createComponent(CreateComponent);

      // check
      expect(titleService.setTitle).toHaveBeenCalledWith(title);
    });
  });

  describe('ogmInputChangeHandler', () => {
    it('should set ogm value to undefined and call next on ogm$ stream with init numberFormat', () => {
      // prepare
      jest.spyOn(component.ogm, 'set');

      const ogmInputChange: Ogm = { ogm: OGM_EMPTY, isValid: null };

      // act
      component.handleOgmChange(ogmInputChange);

      // check
      expect(component.ogm.set).toHaveBeenCalledWith({ number: null, numberFormat: OGM_EMPTY });
    });

    it('should set ogm with the correct data and call next on ogm$ stream with the numberFormat', () => {
      // prepare
      const ogm: OgmData = { number: '120000000002', numberFormat: '+++120/0000/00002+++' };
      jest.spyOn(component.ogm, 'set');

      const ogmInputChange: Ogm = { ogm: ogm.numberFormat, isValid: true };

      // act
      component.handleOgmChange(ogmInputChange);

      // check
      expect(component.ogm.set).toHaveBeenCalledWith({ number: '120000000002', numberFormat: '+++120/0000/00002+++' });
    });
  });

  describe('Copy number or OGM', () => {
    beforeEach(() => {
      jest.spyOn(component, 'showSnackbarMessage');
    });

    describe('copyNumberClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy number!';

        // act
        component.copyNumberClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });

    describe('copyOgmClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy OGM!';

        // act
        component.copyOgmClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });
  });
});

const setup = async () => {
  const renderResult = await render(CreateComponent, {
    componentImports: [MockNumberComponent, MockInputComponent, MockControlsComponent, AsyncPipe],
    providers: [{ provide: MatSnackBar, useValue: { open: jest.fn() } }],
  });

  const titleService = TestBed.inject(Title);

  return {
    ...renderResult,
    component: renderResult.fixture.componentInstance,
    titleService,
  };
};
