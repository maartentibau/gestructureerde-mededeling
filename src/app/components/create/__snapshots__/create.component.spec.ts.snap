// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateComponent rendering component should match snapshot 1`] = `
<ogm-create
  ogm={[Function Function]}
>
  <ogm-number>
    <div>
    </div>
    <div>
    </div>
  </ogm-number><ogm-input
    placeholdermessage="vul een getal in van maximaal 10 cijfers"
  >
    <div>
    </div>
    <div>
    </div>
  </ogm-input><ogm-controls>
    <div>
    </div>
    <div>
    </div>
    <div>
    </div>
    <div>
    </div>
  </ogm-controls>
</ogm-create>
`;
